### DISCUSSION TITLE
Improving memory efficiency

### DISCUSSION THREAD

**Discussion by seanlinsley** [activeadmin/activeadmin#7859] [Ideas]
How can we optimize ActiveAdmin's memory usage? I'd like to discuss that here, ideally with multiple PRs created as a result of that discussion.

I discovered with the rack-mini-profiler and memory_profiler gems that ActiveAdmin creates a lot of objects in memory while rendering a page, just to be thrown away.

You can download the [full report](https://github.com/activeadmin/activeadmin/files/1185129/aa-memory.txt), but here are the highlights:

```
Total allocated: 575396 bytes (8751 objects)
Total retained:  1176 bytes (22 objects)

allocated objects by gem
-----------------------------------
      8751  activeadmin-1.0.0

allocated objects by file
-----------------------------------
      2000  activeadmin-1.0.0/lib/active_admin/resource/naming.rb
       959  activeadmin-1.0.0/lib/active_admin/resource/routes.rb
       883  activeadmin-1.0.0/lib/active_admin/filters/formtastic_addons.rb
       634  activeadmin-1.0.0/lib/active_admin/resource_collection.rb
       491  activeadmin-1.0.0/lib/active_admin/views/components/table_for.rb
       332  activeadmin-1.0.0/lib/active_admin/localizers/resource_localizer.rb
       281  activeadmin-1.0.0/lib/active_admin/views/tabbed_navigation.rb
       259  activeadmin-1.0.0/lib/active_admin/menu_item.rb
       258  activeadmin-1.0.0/lib/active_admin/helpers/settings.rb
       256  activeadmin-1.0.0/lib/active_admin/views/components/scopes.rb

allocated objects by location
-----------------------------------
      1100  activeadmin-1.0.0/lib/active_admin/resource/naming.rb:49
       648  activeadmin-1.0.0/lib/active_admin/filters/formtastic_addons.rb:36
       463  activeadmin-1.0.0/lib/active_admin/resource/naming.rb:20
       330  activeadmin-1.0.0/lib/active_admin/resource_collection.rb:36
       290  activeadmin-1.0.0/lib/active_admin/resource/routes.rb:54
       261  activeadmin-1.0.0/lib/active_admin/resource/routes.rb:102
       206  activeadmin-1.0.0/lib/active_admin/helpers/settings.rb:42
       200  activeadmin-1.0.0/lib/active_admin/resource_collection.rb:38
       165  activeadmin-1.0.0/lib/active_admin/localizers/resource_localizer.rb:31
       135  activeadmin-1.0.0/lib/active_admin/menu_item.rb:67

allocated objects by class
-----------------------------------
      4356  Array
      2850  String
      1250  Hash
       207  Proc
        20  Symbol
        18  ActiveAdmin::Dependency::Matcher
        11  ActiveAdmin::Views::TableFor::Column
        11  Arbre::HTML::ClassList
         9  ActiveAdmin::Dependency::Adapter::Rails
         7  Class
```

This is the query string I used on an index page of my app:
```
?pp=profile-memory&memory_profiler_allow_files=lib/active_admin
```
Refer to the [docs](https://github.com/MiniProfiler/rack-mini-profiler#memory-profiling) for more options.

This is an important issue to fix because it causes unnecessary GC pressure on the thousands of sites that use ActiveAdmin. And if I understand it correctly, retained objects are potential memory leaks.

Memoization is by far the easiest fix when it's possible. `String#freeze` and the [frozen string pragma](https://wyeworks.com/blog/2015/12/1/immutable-strings-in-ruby-2-dot-3) can also help, but they might cause bugs.

An interesting thing in the "Allocated String Report" section is that strings are initialized many times for other models that aren't related to the model in question, and would only be used to render the menu bar.

**Comment by seanlinsley** [comment]
Here are the highlights when including code from all sources:
```
allocated memory by gem
-----------------------------------
  13092602  sprockets-3.7.1
   7566447  ruby-2.4.1/lib
   7153754  rails-8d67a4302042
   1401828  ransack-8ca8c95c6543
   1223667  binding_of_caller-0.7.2
    573972  activeadmin-1.0.0
    417552  i18n-0.8.1
    351254  arbre-1.1.1
    319584  formtastic-3.1.5
    157072  better_errors-2.1.1

allocated memory by class
-----------------------------------
  18175137  String
   4292544  Array
   3941176  Hash
   1553192  MatchData
   1442664  Sprockets::URITar
    867536  RubyVM::InstructionSequence
    748320  Proc
    309197  Regexp
    245560  Set
    172471  ActiveSupport::SafeBuffer

allocated objects by gem
-----------------------------------
    220045  sprockets-3.7.1
     72587  rails-8d67a4302042
     46319  ruby-2.4.1/lib
     17700  ransack-8ca8c95c6543
      8727  activeadmin-1.0.0
      6433  arbre-1.1.1
      3637  formtastic-3.1.5
      3546  i18n-0.8.1
      3240  binding_of_caller-0.7.2
      2994  arel-7.1.4
```

As you can see, Ransack and Arbre are also big offenders.

**Comment by varyonic** [comment]
I started looking at using StringIO buffers instead of String for Arbre a few months ago, you can see my branch at [varyonic/arbre/output_buffer](https://github.com/varyonic/arbre/commits/output_buffer)  Feedback is welcome, but I shelved the effort because I think Arbre has a [bigger problem](https://github.com/activeadmin/arbre/pull/64#issuecomment-288256707) than memory/performance

**Comment by seanlinsley** [comment]
I wouldn't disagree that Arbre has bigger problems, but have they been documented somewhere?

**Comment by varyonic** [comment]
Updated with link above, I'm referring to Arbre interaction with ActionView.  Anyway, using StringIO is an issue worth persuing for Arbre.
